router filter: implement hedge_on_per_try_timeout.

mpuncel opened the pull request:
Implements the hedge_on_per_try_timeout option in HedgePolicy config
which instructs the router filter to not cancel requests that hit the
per try timeout before sending a retry. This means the router must be
able to manage multiple simultaneous upstream requests and handle
deciding which response "wins" and canceling any other in flight
requests.

Finishes #5841.

Description: Implements request hedging (racing) in response to a per try timeout in the router filter.
Risk Level: High
Testing: Unit/Integration tests
Docs Changes: Added a request hedging section to the HTTP routing arch overview in addition to proto-level docs.
Release Notes: updated
Fixes #5841

mpuncel commented on source/common/router/retry_state_impl.cc:
```diff
@@ -193,6 +193,13 @@ RetryStatus RetryStateImpl::shouldRetryReset(Http::StreamResetReason reset_reaso
+RetryStatus RetryStateImpl::shouldHedgeRetryPerTryTimeout(DoRetryCallback callback) {
+  // A hedged retry on per try timeout is always retried if there are retries
+  // left. NOTE: this is different than non-hedged per try timeouts which are only retried
+  // if RETRY_ON_5XX or RETRY_ON_GATEWAY_ERROR
```
This is one place I wasn't sure if this is the right thing to do. There was some discussion on #5841 I'm not caught up on yet

alyssawilk replied:
I will say at first pass this doesn't make sense to folks unfamilar with the retry policies - not sure if it can be made more clear?

mpuncel commented on source/common/router/router.cc:
```diff
@@ -423,20 +442,27 @@ Http::FilterDataStatus Filter::decodeData(Buffer::Instance& data, bool end_strea
+      upstream_requests_[i]->encodeData(copy, end_stream);
+
+      if (i == 0) {
+      }
```
removing

mpuncel commented on source/common/router/router.cc:
```diff
@@ -423,20 +442,27 @@ Http::FilterDataStatus Filter::decodeData(Buffer::Instance& data, bool end_strea
     // buffer limit we give up on retries and buffering. We must buffer using addDecodedData()
     // so that all buffered data is available by the time we do request complete processing and
     // potentially shadow.
     callbacks_->addDecodedData(data, true);
```
I'm assuming we only want to call this once rather than per upstream request

mattklein123 replied:
Yes, definitely.

mpuncel commented on source/common/router/router.cc:
```diff
@@ -502,7 +534,15 @@ void Filter::onRequestComplete() {
 
   // Possible that we got an immediate reset.
-  if (upstream_request_) {
+  bool any_upstreams = false;
```
This logic is a bit funky and I'm not sure what the original intention of the guard against upstream reset is for shadowing, in particular the "riskier" part. Could use some context

mpuncel commented on source/common/router/router.cc:
```diff
@@ -515,44 +555,119 @@ void Filter::onRequestComplete() {
+  // in outlier detection.
+  // TODO(mpuncel) is it weird to have a pretend response code here? we might
+  // get a 200 back from this request later.
+  updateOutlierDetection(timeout_response_code_, upstream_request);
```
could use feedback on this. I think it's good for the outlier detection to know that this host timed out, but it's weird that we might save 2 HTTP codes in for a single request (504 after soft timeout, and then possible 5xx later on).

snowp replied:
Given that outlier detection is about tracking responses from individual hosts I think tracking both is fine

mpuncel commented on source/common/router/router.cc:
```diff
@@ -568,32 +683,15 @@ void Filter::onUpstreamAbort(Http::Code code, StreamInfo::ResponseFlag response_
   // If we have not yet sent anything downstream, send a response with an appropriate status code.
   // Otherwise just reset the ongoing response.
   if (downstream_response_started_) {
-    if (upstream_request_ != nullptr && upstream_request_->grpc_rq_success_deferred_) {
```
had to decompose `onUpstreamAbort` a bit more because some of it should run per-upstream and some is once. Only affects global timeout where we might be canceling multiple upstream requests at once

mpuncel commented on source/common/router/router.cc:
```diff
@@ -637,23 +726,25 @@ bool Filter::maybeRetryReset(Http::StreamResetReason reset_reason) {
-void Filter::onUpstreamReset(Http::StreamResetReason reset_reason) {
-  ASSERT(upstream_request_);
+void Filter::onUpstreamReset(Http::StreamResetReason reset_reason,
+                             UpstreamRequest* upstream_request) {
```
Let me know if it's weird to be using raw pointers here, alternatively could make UpstreamRequest a shared pointer (it's a unique pointer)

snowp replied:
You can probably pass a ref instead since it won't ever be null

mpuncel commented on source/common/router/router.cc:
```diff
@@ -677,63 +768,91 @@ Filter::streamResetReasonToResponseFlag(Http::StreamResetReason reset_reason) {
-    const RetryStatus retry_status =
+  // Check if this upstream request was already retried, for instance after
+  // hitting a per try timeout. Don't retry it if we already have.
+  if (retry_state_ && !upstream_request->retried_) {
```
realized I can use another unit test that makes sure we don't retry the same upstream request twice (once for per try timeout and then again for a 5xx for example).

mpuncel commented:
I'd like some integration tests for this as well, but will probably wait until Matt is done with the one he's writing so I can use it as a guide.

mpuncel commented:
working through the build failures and will add more tests today

mpuncel commented on include/envoy/router/router.h:
```diff
@@ -240,6 +240,16 @@ class RetryState {
+   * @param response_headers supplies the response headers.
+   * @return bool true if a retry would be warranted based on the retry policy.
+   */
+  virtual bool wouldRetryFromHeaders(const Http::HeaderMap& response_headers) PURE;
```
if exposing this isn't desirable, there are some alternatives we could discuss. Right now i'm making sure we never send more than 1 retry for the same upstream request, which is why I needed an API for "i'm not going to retry this, but should I if I was?". The reason I wanted to do that is so we don't send any more retries than in the non-hedging case, the only difference is whether we reset the original request or not.

snowp replied:
I think exposing this is fine assuming we want the behavior of "accept first response that wouldn't be retried". The other option "accept first response that isn't retried" is simpler but perhaps less ideal as you might be able to wait just a short while for a "better" response.

mpuncel commented on source/common/router/router.h:
```diff
@@ -371,32 +394,41 @@ class Filter : Logger::Loggable<Logger::Id::router>,
+                         UpstreamRequest* upstream_request, bool end_stream);
+  void onUpstreamData(Buffer::Instance& data, UpstreamRequest* upstream_request, bool end_stream);
+  void onUpstreamTrailers(Http::HeaderMapPtr&& trailers, UpstreamRequest* upstream_request);
   void onUpstreamMetadata(Http::MetadataMapPtr&& metadata_map);
```
not sure if anything needs to change with this function. right now all upstream metadata gets proxied downstream

mpuncel commented:
This is no longer WIP and ready for comment

snowp commented on source/common/router/router.h:
```diff
@@ -285,6 +301,9 @@ class Filter : Logger::Loggable<Logger::Id::router>,
       upstream_host_ = host;
       parent_.callbacks_->streamInfo().onUpstreamHostSelected(host);
+      if (parent_.retry_state_ && host) {
+        parent_.retry_state_->onHostAttempted(host);
```
Is this the only place we call this now?

mpuncel replied:
yeah

snowp commented on source/common/router/router.cc:
```diff
@@ -380,8 +396,10 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
+  UpstreamRequestPtr upstream_request = std::make_unique<UpstreamRequest>(*this, *conn_pool);
+  UpstreamRequest* upstream_request_ptr = upstream_request.get();
+  upstream_request->moveIntoList(std::move(upstream_request), upstream_requests_);
+  upstream_request_ptr->encodeHeaders(end_stream);
```
Instead of using a temp variable you could just do `upstream_requests_.front()->encodeHeaders(end_stream);`

snowp commented on source/common/router/router.cc:
```diff
@@ -423,20 +441,25 @@ Http::FilterDataStatus Filter::decodeData(Buffer::Instance& data, bool end_strea
-    // since it's all moves from here on.
-    Buffer::OwnedImpl copy(data);
-    upstream_request_->encodeData(copy, end_stream);
+  for (auto it = upstream_requests_.cbegin(); it != upstream_requests_.cend(); it++) {
```
Might be cleaner here to use a for each loop

snowp commented on source/common/router/router.cc:
```diff
@@ -423,20 +441,25 @@ Http::FilterDataStatus Filter::decodeData(Buffer::Instance& data, bool end_strea
+  for (auto it = upstream_requests_.cbegin(); it != upstream_requests_.cend(); it++) {
+    UpstreamRequest* upstream_request = it->get();
+    if (buffering) {
+      // If we are going to buffer for retries or shadowing, we need to make a copy before encoding
```
Do we need to make a copy in the non-buffering case too now that we're passing `data` to multiple `encodeData`?

mpuncel replied:
yeah makes sense, should I avoid copying if there is 1 upstream request and not buffering?

snowp commented on source/common/router/router.cc:
```diff
@@ -449,7 +472,9 @@ Http::FilterDataStatus Filter::decodeData(Buffer::Instance& data, bool end_strea
   ENVOY_STREAM_LOG(debug, "router decoding trailers:\n{}", *callbacks_, trailers);
   downstream_trailers_ = &trailers;
-  upstream_request_->encodeTrailers(trailers);
+  for (auto it = upstream_requests_.cbegin(); it != upstream_requests_.cend(); it++) {
```
same comment here, for each loop might be nicer

snowp commented on source/common/router/router.cc:
```diff
@@ -514,45 +540,112 @@ void Filter::onRequestComplete() {
-  updateOutlierDetection(timeout_response_code_);
+// Called when the per try timeout is hit but we didn't reset the request
+// (hedge_on_per_try_timeout enabled).
+void Filter::onSoftPerTryTimeout(UpstreamRequest* upstream_request) {
```
Since this is never null from what I can tell it would probably be better to pass this by reference, same with other places where you pass a pointer

snowp commented on source/common/router/router.cc:
```diff
@@ -638,23 +705,25 @@ bool Filter::maybeRetryReset(Http::StreamResetReason reset_reason) {
     return;
   }
 
+  bool dropped = reset_reason == Http::StreamResetReason::Overflow;
```
nit: const

snowp replied:
bump

snowp commented on source/common/router/router.cc:
```diff
@@ -679,84 +748,132 @@ Filter::streamResetReasonToResponseFlag(Http::StreamResetReason reset_reason) {
+      if (!upstream_request_tmp->encode_complete_ || !upstream_request_tmp->decode_complete_) {
+        upstream_request_tmp->resetStream();
+        if (upstream_request_tmp->upstream_host_) {
+          upstream_request_tmp->upstream_host_->stats().rq_hedge_abandoned_.inc();
```
We should probably also have a cluster level stat for this

mpuncel replied:
i have one at L794 right?

snowp commented on source/common/router/router.cc:
```diff
@@ -679,84 +748,132 @@ Filter::streamResetReasonToResponseFlag(Http::StreamResetReason reset_reason) {
+
+void Filter::resetOtherUpstreams(UpstreamRequest* upstream_request) {
+  UpstreamRequest* upstream_request_tmp;
+  for (auto it = upstream_requests_.cbegin(); it != upstream_requests_.cend(); it++) {
```
for each loop

snowp commented on source/common/router/router.cc:
```diff
@@ -679,84 +748,132 @@ Filter::streamResetReasonToResponseFlag(Http::StreamResetReason reset_reason) {
+  for (auto it = upstream_requests_.cbegin(); it != upstream_requests_.cend(); it++) {
+    upstream_request_tmp = it->get();
+    if (upstream_request_tmp != upstream_request) {
+      if (!upstream_request_tmp->encode_complete_ || !upstream_request_tmp->decode_complete_) {
```
Is this just implicitly checking if the request is not the winner? Is there a more explicit way? If not, add a comment?

snowp replied:
bump

mpuncel replied:
this was unnecessary because it's checked in resetStream, so removed it

snowp commented on source/common/router/router.cc:
```diff
@@ -679,84 +748,132 @@ Filter::streamResetReasonToResponseFlag(Http::StreamResetReason reset_reason) {
+      // We already retried this request (presumably for a per try timeout) so
+      // we definitely won't retry it again. Check if we would have retried it
+      // if we could.
+      could_not_retry = retry_state_->wouldRetryFromHeaders(*headers);
```
If this variable says that we wanted to retry but couldn't, why is it set to true when we hit the retry count limit?

snowp replied:
bump

mpuncel replied:
it means like "request was bad but won't/can't retry it". That could be either because we already retried it w/ hedging or because we hit the limit

snowp commented on source/common/router/router.cc:
```diff
@@ -679,84 +748,132 @@ Filter::streamResetReasonToResponseFlag(Http::StreamResetReason reset_reason) {
-      callbacks_->streamInfo().setResponseFlag(
-          StreamInfo::ResponseFlag::UpstreamRetryLimitExceeded);
+    if (upstream_request->retried_) {
+      // We already retried this request (presumably for a per try timeout) so
```
Can you make sure this behavior is documented?

snowp commented on source/common/router/router.cc:
```diff
@@ -679,84 +748,132 @@ Filter::streamResetReasonToResponseFlag(Http::StreamResetReason reset_reason) {
+      // if we could.
+      could_not_retry = retry_state_->wouldRetryFromHeaders(*headers);
+    } else {
+      RetryStatus retry_status =
```
const

snowp commented on source/common/router/router.cc:
```diff
@@ -935,6 +1060,8 @@ bool Filter::setupRedirect(const Http::HeaderMap& headers) {
 void Filter::doRetry() {
   is_retry_ = true;
   attempt_count_++;
+  ASSERT(pending_retries_ > 0);
```
just make `pending_retries_` unsigned to avoid it going below 0

mpuncel replied:
my intention here was to blow up instead of overflow, but i guess i could `assert > 0` before decrementing

mpuncel replied:
oh i do do that lol

snowp commented on source/common/router/router.cc:
```diff
@@ -935,6 +1060,8 @@ bool Filter::setupRedirect(const Http::HeaderMap& headers) {
   is_retry_ = true;
   attempt_count_++;
+  ASSERT(pending_retries_ > 0);
+  pending_retries_--;
```
How can you have more than one pending retry when there's only one retry callback tracked at any point?

mpuncel replied:
i can't, but when implementing the inititial requests part I might.

snowp commented on source/common/router/router.cc:
```diff
@@ -947,28 +1074,42 @@ void Filter::doRetry() {
+  UpstreamRequestPtr upstream_request = std::make_unique<UpstreamRequest>(*this, *conn_pool);
+  UpstreamRequest* upstream_request_ptr = upstream_request.get();
+  upstream_request->moveIntoList(std::move(upstream_request), upstream_requests_);
+  upstream_request_ptr->encodeHeaders(!callbacks_->decodingBuffer() && !downstream_trailers_);
```
use `upstream_requests_.front()`

snowp commented on docs/root/intro/arch_overview/http_routing.rst:
```diff
@@ -87,6 +88,16 @@ headers <config_http_filters_router_headers_consumed>`. The following configurat
 
+.. _arch_overview_http_routing_hedging:
+
+Request Hedging
```
Can you add some more detail on how this interacts with retries etc.?

mattklein123 replied:
+1, also can you be more explicit on how hedging on per_try_timeout effectively disables the per_try_timeout and let's the original request continue on? That wasn't clear to me until I started looking through the implementation.

mpuncel replied:
updated the wording here, I think it's more clear now but let me know if it can use more work

snowp commented on source/common/router/router.cc:
```diff
@@ -380,8 +396,9 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
-  upstream_request_ = std::make_unique<UpstreamRequest>(*this, *conn_pool);
-  upstream_request_->encodeHeaders(end_stream);
+  UpstreamRequestPtr upstream_request = std::make_unique<UpstreamRequest>(*this, *conn_pool);
+  upstream_request->moveIntoList(std::move(upstream_request), upstream_requests_);
```
Seems like we're now keeping track of all the upstream requests even in the non-hedging case, we probably want to remove them from the list when we're done with them to ensure that we're not changing the current, non-hedging behavior.

alyssawilk replied:
Totally optional and quite possibly overkill, but it might be worth making some of the single->list changes in their own PR where there's no functional change, to make it more clear what the deltas are.

mpuncel replied:
I gave this suggestion a try and it seems like a good standalone PR: https://github.com/envoyproxy/envoy/pull/6540/files

snowp commented on source/common/router/router.cc:
```diff
@@ -423,20 +440,21 @@ Http::FilterDataStatus Filter::decodeData(Buffer::Instance& data, bool end_strea
+    // shadowing/retries.
     Buffer::OwnedImpl copy(data);
-    upstream_request_->encodeData(copy, end_stream);
+    upstream_request->encodeData(copy, end_stream);
```
We probably want to avoid a copy here in the non-hedging case

snowp commented on source/common/router/router.cc:
```diff
@@ -514,45 +535,111 @@ void Filter::onRequestComplete() {
+  updateOutlierDetection(timeout_response_code_, upstream_request);
+  upstream_request->outlier_detection_timeout_recorded_ = true;
+
+  Upstream::HostDescriptionConstSharedPtr upstream_host = upstream_request->upstream_host_;
```
Is this used?

snowp commented on api/envoy/api/v2/route/route.proto:
```diff
@@ -859,22 +857,27 @@ message RetryPolicy {
-  // timed out request should be retried. Defaults to false.
+  // timed out request should be retried.
+  // Once a timed out request is retried due to per try timeout, it will not be
+  // retried again even if the returned response headers would otherwise b
```
incomplete sentence

snowp commented on docs/root/intro/arch_overview/http_routing.rst:
```diff
@@ -87,6 +88,18 @@ headers <config_http_filters_router_headers_consumed>`. The following configurat
+---------------
+
+Envoy supports request hedging via specifying a :ref:`hedge policy <envoy_api_msg_route.HedgePolicy>`. This means that Envoy
+will race multiple simultaneous upstream requests and return the first valid response to the downstream.
```
Maybe note that "valid" is based on the retry policy?

mpuncel commented:
tried this build out again, and hit an assert in `addDownstreamWatermarkCallbacks` because it gets called multiple times. will need to take a look at that

mattklein123 commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -1477,16 +1477,16 @@ bool ConnectionManagerImpl::ActiveStream::verbose() const {
   ++high_watermark_count_;
-  if (watermark_callbacks_) {
-    watermark_callbacks_->onAboveWriteBufferHighWatermark();
+  for (auto watermark_callbacks : watermark_callbacks_) {
```
Can you add explicit HCM unit tests for these changes? cc @alyssawilk to also look through this part.

alyssawilk replied:
Also E2E tests if you can.  Some of the large body integration test with low buffer limits are worth checking out.

mpuncel replied:
Any particular cases you'd like to see covered? I'm not very familiar with the buffer watermark lifecycle.

mpuncel commented:
@mattklein123 are you able to take a look to make sure you like the direction of this change? If we agree on the behavior here I can work on adding integration tests

mattklein123 commented:
@mpuncel sorry we raced comments. At a high level it looks sane to me, but it's a super complicated change and extremely high risk so I need time to look through. I think it's fine to start with integration tests, since I doubt the user visible behavior will change? WDYT?

mpuncel commented:
sounds good to me! will address your comments from today and work on integration tests

alyssawilk commented on docs/root/intro/arch_overview/http_routing.rst:
```diff
@@ -87,6 +88,21 @@ headers <config_http_filters_router_headers_consumed>`. The following configurat
+Envoy supports request hedging via specifying a :ref:`hedge policy
+<envoy_api_msg_route.HedgePolicy>`. This means that Envoy will race multiple
+simultaneous upstream requests and return the first valid response to the
+downstream according to retry policy.
```
Reading this I couldn't tell if we wait for full response, or first successful-enough response headers.  Looks from code like we do the latter.  Could we clarify the docs on this point?

alyssawilk commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -1801,19 +1801,20 @@ void ConnectionManagerImpl::ActiveStreamDecoderFilter::
 
 void ConnectionManagerImpl::ActiveStreamDecoderFilter::addDownstreamWatermarkCallbacks(
     DownstreamWatermarkCallbacks& watermark_callbacks) {
-  // This is called exactly once per stream, by the router filter.
```
"Once per stream" could be updated for clarity.

Before, upstream and downstream streams functionally had 1:1 correspondence (well, to be really picky, one a time for retries).  Once per upstream-stream?

Also I'd be concerned that if you had an upstream which was timing out, it wouldn't be acking data, so your hedge would fail.  This is somewhat mitigated by the fact retries currently work for fully buffered requests, but could cause your hedged request to fail due to the outstanding stream, a problematic design flaw which might be worth documenting.  Alternately if we can do more to detect problematic configs, say one with a larger listener perConnectionBufferLimitBytes than cluster perConnectionBufferLimitBytes with hedging on.

mpuncel replied:
Ah interesting, I hadn't thought of that. Do you mind explaining a bit more why the case where listener buffer size is larger than cluster connection buffer size is particularly problematic? Not sure I get that part.

alyssawilk commented on source/common/router/router.cc:
```diff
@@ -423,20 +441,25 @@ Http::FilterDataStatus Filter::decodeData(Buffer::Instance& data, bool end_strea
+      Buffer::OwnedImpl copy(data);
+      upstream_request->encodeData(copy, end_stream);
+    } else {
+      upstream_request->encodeData(data, end_stream);
```
AFIK on this path the common case is void Filter::UpstreamRequest::encodeData -> request_encoder_->encodeData which is allowed to move or drain data.   I'm not sure both request bodies will make it upstream.

I strongly encourage some integration tests here.  Even if the internals of router/hcm change I think the overall data flow between Envoy and upstream should remain the same, so I don't think you'll end up with any time wasted!

mattklein123 replied:
+1 I don't think this will work if there are multiple upstreams without copying all but on of them. Is this path tested?

mpuncel replied:
I think I should revert this code and instead do an assert that `upstream_requests_.size() == 1`

snowp commented:
Gonna mark this as no stalebot since Mike is out this and next week

mattklein123 commented:
Please merge master and move release note to 1.11.0 and let's get this reviewed and merged!

/wait

mpuncel commented:
merged master and moved release note, will start working through Alyssa's feedback

mpuncel commented:
added some integration tests, and a `x-envoy-hedge-on-per-try-timeout` header (mostly to make the integration test easier to set up but is also a useful feature).

I expect the test I added to fail until https://github.com/envoyproxy/envoy/pull/6526 is merged in because the last retry is the one that succeeds, but currently we see a retry overflow value.

mpuncel commented:
I think this is in a pretty good state. I didn't write any unit tests for the conn manager changes because I think the integration tests cover it pretty well. I have cases there for large request with small upstream buffer size, and large response with small downstream buffer size. There is no retrying if the full request can't be buffered so that means there is no hedging either.

mpuncel commented:
/retest

mpuncel commented:
/retest

mpuncel commented:
/retest

mpuncel commented:
/retest

mpuncel commented:
/retest

mpuncel commented:
oops github was telling me my comments were failing but they were actually going through

mattklein123 commented:
Will give this a detailed review once we merge the breakout PR.

/wait

mpuncel commented:
unfortunately I can't seem to reproduce that ipv6 flake anywhere but on circleci so i'm not sure what's going on

mpuncel commented:
I merged master and pushed that commit up, but I still need to do some more work on the watermark stuff, so no need to review yet.

mpuncel commented on source/common/router/router.h:
```diff
@@ -303,12 +321,21 @@ class Filter : Logger::Loggable<Logger::Id::router>,
+      // disabling reads will not slow down other upstream requests. If we've
+      // already seen the full downstream requst (downstream_end_stream_) then
+      // disabling reads is a no-op.
+      ASSERT(parent_.upstream_requests_.size() == 1 || downstream_end_stream_);
```
This assert isn't always true until https://github.com/envoyproxy/envoy/pull/6643/files is merged because we might hit a per try timeout before `downstream_end_stream_` is true

alyssawilk commented:
I definitely think we should sort out the watermark issues (as a separate PR!) and land them before we unhide any hedge config.  Happy to chat about them with both of you now that I'm back.

mpuncel commented:
@alyssawilk I think the PR is in a good spot with the watermark stuff (although I believe i'm still missing some of the regression tests you asked for w.r.t. the for loop over the list of registered callbacks).

There are a couple of observations that made things simpler:

1) for the "request too big" case, there should only be one in flight upstream request. This assumption holds true because a) we can only have multiple upstream requests if there was a per try timeout and b) (after https://github.com/envoyproxy/envoy/pull/6643/files) a per try timeout can only occur after we've seen the whole downstream request. That means I can comfortably pause downstream reads and know I'm not slowing down other requests. 

2) for the "response too big" case, I'm either actively writing data downstream after having selected a "winning" upstream request and resetting the others, or there is no "winning" upstream request at all and i'm not writing data downstream. In either case it's okay for me to read disable upstream requests and wait for a below watermark callback.

alyssawilk commented:
1) you're totally right that #6643 helps you out here, insofar as you simply won't retry or hedge at all in cases where the backend is behaving badly and doesn't drain the tcp socket buffer.  It's kind of a bummer, because it means that both hedging and retries may simply not work for large requests and new connections because if the TCP window is not warmed, the buffer will fill, reading will stop, and the data will simply never be read.  Looks like setUpRetry already fast failed in that case so it's no worse, it just means there' a whole class of failures we don't retry or hedge for.  The other thing required for this to work was correct unwinding, which landed in  #6578.  Now the codec will which will readEnable for subsequent HTTP/1 requests after the hedge is successful. For H2 that's already sorted out onStreamClose.

for 2) I think you're good given you reset the other upstreams once you have headers. I think you can update hedge docs a bit to make it more clear (rather than good response, good response headers etc.) but watermarks shouldn't be a problem here.  yay!

mpuncel commented:
looks like my PR made cluster memory size go above the test threshold?
```
[ RUN      ] IpVersions/ClusterMemoryTestRunner.MemoryLargeClusterSizeWithStats/IPv4
test/integration/stats_integration_test.cc:199: Failure
Expected: (m_per_cluster) < (59600), actual: 59848 vs 59600
```
Could that be because of the stat I added?

alyssawilk commented:
yeah, sounds plausible.  if I recall correctly @cmluciano added this test but @jmarantz put a bunch of work in to making the stats scale better memory wise.  Unfortunately I don't think we have a great way of handling stats for off-by-default features, where we want the stat if it's configured and we don't want to use memory for all the stats for all the features for all the clusters.  Maybe Josh has thoughts here?

mattklein123 commented:
@mpuncel @alyssawilk sorry catching up. Is this ready to review again or are we still waiting on changes?

jmarantz commented:
There's at least one other case I reviewed (forgot exactly where) of having some extra stats enabled by flag. In that case the stats made sense by default, but you only get them if you ask for them by turning on a feature.

In this case it seems easy enough to refrain from initializing a stat when the  feature it's measuring is turned off.

jmarantz commented:
BTW the stat memory-scaling work is still ongoing; the biggest gains are pending (including one waiting for review in #6161).

mpuncel commented:
@mattklein123 it's ready for a look (aside from this cluster memory issue)!

mpuncel commented:
I'm not really sure what my options here are for selectively enabling the stat. Any HTTP request could enable the feature with headers.

mattklein123 commented on docs/root/intro/version_history.rst:
```diff
@@ -3,6 +3,7 @@ Version history
 
 1.11.0 (Pending)
 ================
+* router: added ability to issue a hedged retry in response to a per try timeout via a :ref:`hedge policy <envoy_api_msg_route.HedgePolicy>`.
```
nit: alpha order

mattklein123 commented on docs/root/operations/admin.rst:
```diff
@@ -84,6 +84,7 @@ modify different aspects of the server:
       cx_total, Counter, Total connections
       cx_active, Gauge, Total active connections
       cx_connect_fail, Counter, Total connection failures
+      rq_hedge_abandoned, Counter, Total hedged requests that were canceled and abandoned due to accepting another response.
```
Should we also have a counter on hedged request total? WDYT?

mpuncel replied:
will add one

mattklein123 commented on source/common/router/router.h:
```diff
@@ -89,6 +94,16 @@ class FilterUtility {
+   *        initial request should be sent
+   * @return HedgingParams the final parameters to use for request hedging
+   */
+  static HedgingParams finalHedgingParams(const RouteEntry& route, Http::HeaderMap& request_headers,
```
nit: missing request_headers in docs

mattklein123 commented on source/common/router/router.h:
```diff
@@ -303,12 +321,21 @@ class Filter : Logger::Loggable<Logger::Id::router>,
 
     void disableDataFromDownstream() {
-      ASSERT(parent_.upstream_requests_.size() == 1);
+      // If there is only one upstream request, we can be assured that
```
Can you add additional text here and below and how we know this assert is true in the case of hedging? It might be easier to put this assert in a helper function where the comment can leave, but nbd either way and up to you.

mattklein123 commented on test/integration/http_timeout_integration_test.cc:
```diff
@@ -95,4 +95,145 @@ TEST_P(HttpTimeoutIntegrationTest, PerTryTimeout) {
+// try but slightly less than the global timeout. We then have the first
+// upstream request return headers and expect those to be returned downstream
+// (which proves the request was not canceled when the timeout was hit).
+TEST_P(HttpTimeoutIntegrationTest, HedgedPerTryTimeout) {
```
Can you make sure to run these integration tests with runs_per_test set to 1000+ to look for flakes?

mpuncel replied:
ran it with 1000 runs per test and didn't get any flakes

mattklein123 commented on source/common/router/router.cc:
```diff
@@ -193,6 +193,33 @@ FilterUtility::finalTimeout(const RouteEntry& route, Http::HeaderMap& request_he
+                                                               Http::HeaderMap& request_headers,
+                                                               uint64_t random_value) {
+  HedgingParams hedgingParams;
+  hedgingParams.initial_requests_ = route.hedgePolicy().initialRequests();
```
This is not implemented yet, right? (It's marked as such in the docs). If so, can we remove this code for now from the struct and here?

mattklein123 commented on source/common/router/router.cc:
```diff
@@ -541,37 +578,85 @@ void Filter::onRequestComplete() {
+void Filter::onSoftPerTryTimeout(UpstreamRequest& upstream_request) {
+  // Even though we didn't cancel the request yet we still want to track it
+  // in outlier detection.
+  // TODO(mpuncel) is it weird to have a pretend response code here? we might
```
I think it's probably fine, given that we have seen what we think is an outlier from this host, even if it shows up late later. We don't charge it as a success later, right (per the boolean)? If you agree I would just remove this TODO.

mpuncel replied:
sounds good to me

mattklein123 commented on source/common/router/router.cc:
```diff
@@ -541,37 +578,85 @@ void Filter::onRequestComplete() {
   updateOutlierDetection(timeout_response_code_, upstream_request);
 
   if (maybeRetryReset(Http::StreamResetReason::LocalReset, upstream_request)) {
+    upstream_request.removeFromList(upstream_requests_);
```
It scares me a little that we are doing list removals in a return case as specific as this. Is there somewhere more central we could do this? I'm not sure what the options would be.

mpuncel commented:
@mattklein123 @alyssawilk ping for review and advice on what to do about cluster memory. One option would be to remove the new stat for now and deal with it in another PR

mattklein123 commented:
> @mattklein123 @alyssawilk ping for review and advice on what to do about cluster memory. One option would be to remove the new stat for now and deal with it in another PR

If you want to add the stats in a follow up that's fine with me, but I do think we should add them (and I mentioned another one we might want to add).

jmarantz commented:
RE "I'm not really sure what my options here are for selectively enabling the stat. Any HTTP request could enable the feature with headers." -- one option is that you could only create that stat when someone enables the feature. Another option is to determine that the stat/feature is sufficiently important that we should always have the stat for it even if the feature is disabled. I'm OK with any of that; my position is that I just want to be conscious of the memory consumed.

mattklein123 commented:
> RE "I'm not really sure what my options here are for selectively enabling the stat. Any HTTP request could enable the feature with headers." -- one option is that you could only create that stat when someone enables the feature. Another option is to determine that the stat/feature is sufficiently important that we should always have the stat for it even if the feature is disabled. I'm OK with any of that; my position is that I just want to be conscious of the memory consumed.

It might be sufficient in this case to make it a router stat vs. cluster stat. The other option is to only allow header override if the feature is configured in some way, then create the stats in a new sub-scope. Personally I think I prefer the latter option?

mpuncel commented:
pushed a new commit addressing *most* of the last round of feedback. Will mark the relevant items resolved, there are a couple that will require more thought on my end

mpuncel commented:
FYI I'm going to do a wider internal release of this branch to Square today

mpuncel commented:
I am going to do another pass to make sure the coverage is good, probably I can test more permutations of bad headers, resets, and per try timeouts in the integration tests

mpuncel commented:
By the way this is what the upstream access logs look like when making a call that takes 2s in the server with a per try timeout of 100ms:

```
[2019-04-26T19:27:16.298Z] "GET /sleep/2s HTTP/2" 200 -/- UT 0 0 2002 2002 "curl/7.29.0" "3280f316-acc0-43bc-abff-7895bda99574" "envoy-test.slug.gns.square" "-" "10.1.3.11:32351"
[2019-04-26T19:27:16.619Z] "GET /sleep/2s HTTP/2" 0 -/- UT 0 0 1681 - "curl/7.29.0" "3280f316-acc0-43bc-abff-7895bda99574" "envoy-test.slug.gns.square" "-" "10.1.3.11:32351"
[2019-04-26T19:27:16.957Z] "GET /sleep/2s HTTP/2" 0 -/- UT 0 0 1344 - "curl/7.29.0" "3280f316-acc0-43bc-abff-7895bda99574" "envoy-test.slug.gns.square" "-" "10.1.3.11:32351"
[2019-04-26T19:27:17.331Z] "GET /sleep/2s HTTP/2" 0 -/- UT 0 0 969 - "curl/7.29.0" "3280f316-acc0-43bc-abff-7895bda99574" "envoy-test.slug.gns.square" "-" "10.1.3.11:32351"
[2019-04-26T19:27:17.711Z] "GET /sleep/2s HTTP/2" 0 -/- UT 0 0 590 - "curl/7.29.0" "3280f316-acc0-43bc-abff-7895bda99574" "envoy-test.slug.gns.square" "-" "10.1.3.11:32351"
```

mpuncel commented:
I've removed the hedging stats for now because I'm wary of adding to this PR. Will create an issue to follow up and add the stats in a way that doesn't increase cluster memory without being enabled explicitly.

mattklein123 commented:
@mpuncel what's the status here, are you still working on changes or ready for another pass?

mpuncel commented:
@mattklein123 I'm ready for another pass. Since you last looked the big changes are:

1) removed the stats to avoid the memory issues. I looked into adding a sub scope but it seems complex enough that I'd rather do it in a followup PR if that's alright with you.
2) `grpc-timeout` and `x-envoy-expected` now get set to the global timeout instead of per try timeout when hedge_on_per_try_timeout is enabled. Without this change upstream servers might time out and negate the feature.
3) Added more cases to the integration tests, now we try 1st request succeeding and 2nd request succeeding for each test case.
4) there's now a cluster option `allow_request_hedging` that needs to be on for hedging to be enabled. Without it the route, virtual host, and header configs are ignored. This is mostly in preparation for the stats followup later to avoid backwards compatibility issues where someone can enable hedging without stats

mattklein123 commented on api/envoy/api/v2/cds.proto:
```diff
@@ -581,6 +581,10 @@ message Cluster {
+
+  // If enabled, allow HTTP/gRPC requests to this cluster to use a hedging
+  // strategy in which multiple upstream requests may be sent simultaneously.
+  bool allow_request_hedging = 39;
```
Why not just put this on the route? It doesn't seem like we need a cluster level config for this? It seems like the route/vhost config could be defined such that if the message is present it turns it on, even though it might not happen without headers?

mpuncel replied:
I did that in response to your previous comment:

> It might be sufficient in this case to make it a router stat vs. cluster stat. The other option is to only allow header override if the feature is configured in some way, then create the stats in a new sub-scope. Personally I think I prefer the latter option?

I could be wrong here, but I was thinking that the decision to allocate memory for cluster stats has to happen when the cluster is initialized, hence why I put this on the cluster. I'm requiring it up front because if I don't in this PR someone might enable it via route and not via the cluster and then be using the feature without stats.

mattklein123 replied:
Hmm yeah fair enough. I guess I was thinking that we could dynamically allocate the stats the first time they are used on the route/cluster, but that might be a bit difficult unless we make them router stats and not cluster stats? It seems not great to have things like this hanging off of the cluster but I can't think of a better option in this case. WDYT about at least putting this inside `core.HttpProtocolOptions`? Any other ideas?

mpuncel replied:
If we don't want this option maybe I should just remove it for now? I think the best way would be to add "lazy" stats that only are initialized when first incremented, then we don't need config options or to use more memory in the common case.

mattklein123 replied:
I'm still not sure how you would add lazy stats without making the cluster code more complicated. It's almost like we want cluster scoped router specific stats, but that is also complicated from a cleanup perspective. @envoyproxy/maintainers any thoughts on this one? @jmarantz in particular for comment?

snowp replied:
It seems to me like we're in a problematic spot where we used to just enable stats by default and that we're now in a situation where we find ourselves adding flags scattered around the code base to control whether to enable features to keep the number of stats under control. Maybe instead there should be be a way to centrally control stats - kinda like the existing `StatsMatcher` except that we default new stats to not being included. That way we could keep stats under control without having to add in feature specific config whenever new stats are added

Given how complicated this stats question is turning out, how do we feel about landing this change without stats and figuring out the best way of handling stats separately? I know we'd love to see this feature land and while stats would be nice we could get by without them for the time being

mattklein123 replied:
I've come to the same conclusion as @snowp. I think we should probably trim down the cluster stats by default with a default exclusion set, and then people can-opt in, and we can document. Yes I think it's fine landing this without stats and then figuring it out. +1

mattklein123 commented:
@mpuncel can you merge master to pick up various CI fixes and I will take a final pass? Thank you.

/wait

mattklein123 commented on source/common/router/router.cc:
```diff
@@ -993,7 +1128,7 @@ void Filter::doRetry() {
   upstream_requests_.front()->encodeHeaders(!callbacks_->decodingBuffer() && !downstream_trailers_);
   // It's possible we got immediately reset.
-  if (upstream_requests_.size() == 1) {
+  if (upstream_requests_.front()) {
```
nit: `!upstream_requests_.empty()` ?

mpuncel commented on source/common/router/router.cc:
```diff
@@ -846,11 +979,12 @@ void Filter::onUpstreamData(Buffer::Instance& data, UpstreamRequest& upstream_re
   }
 
-  callbacks_->encodeData(data, end_stream);
+  if (final_upstream_request_ == &upstream_request) {
```
Looking at this again, I don't think we need this block because when we set `final_upstream_request_` we reset all other upstreams which also removes the callbacks. We might not need `final_upstream_request_` at all actually for that reason

snowp commented on source/common/router/router.cc:
```diff
@@ -1174,11 +1331,9 @@ void Filter::UpstreamRequest::onPerTryTimeout() {
-    }
-    resetStream();
+
+    // Set response flag to UT for now, but it might be overwritten if a
```
`setResponseFlag` ORs together each passed value, subsequent calls won't overwrite previous calls.

snowp commented on source/common/router/router.cc:
```diff
@@ -1286,7 +1442,15 @@ void Filter::UpstreamRequest::clearRequestEncoder() {
+  // There are two states we should get this callback in: 1) the watermark was
+  // hit due to writes from a different filter instance over a shared
+  // downstream connection, or 2) the watermark was hit due to THIS filter
+  // instance due to writing back the "winning" upstream request. In either
```
nit: due to is repeated twice

snowp commented on source/common/router/router.cc:
```diff
@@ -1006,11 +1141,24 @@ void Filter::doRetry() {
 }
 
+uint32_t Filter::numRequestsAwaitingHeaders() {
+  uint32_t ret = 0;
```
you can use `std::count_if` here

snowp commented on test/common/http/conn_manager_impl_test.cc:
```diff
@@ -3143,10 +3143,13 @@ TEST_F(HttpConnectionManagerImplTest, HitFilterWatermarkLimits) {
   buffer_len = encoder_filters_[1]->callbacks_->encodingBuffer()->length();
   EXPECT_CALL(callbacks, onBelowWriteBufferLowWatermark());
-  EXPECT_CALL(callbacks2, onBelowWriteBufferLowWatermark());
+  EXPECT_CALL(callbacks2, onBelowWriteBufferLowWatermark()).Times(0);
```
Can you just remove this line if you don't expect it to be called?

mpuncel replied:
I think I need that to make sure it's _not_ called?

snowp replied:
As long as the mock is not a NiceMock it will fail the test with "uninteresting function called on ..." if you haven't added an expectation for it

There is possibly some value in being visually explicit in this assertion but it's not necessary afaik

mpuncel commented on source/common/router/router.cc:
```diff
@@ -760,31 +862,42 @@ void Filter::onUpstreamHeaders(uint64_t response_code, Http::HeaderMapPtr&& head
+                                           static_cast<Http::Code>(response_code));
+        upstream_host->stats().rq_error_.inc();
+        return;
+      } else if (retry_status == RetryStatus::NoOverflow) {
```
it's enough different that' it'd be difficult. We call a different retry state function, reset the stream on headers but not soft per try timeout, charge http code on headers and not soft per try timeout, and also have the `could_not_retry` logic which tells us whether we're done (e.g. this was the last in-flight attempt) vs need to wait for other attempts to potentially get headers back.

I'll take a look into deduplicating the retry handing `onSoftPerTryTimeout` with `maybeRetryReset` though, those look similar enough.

mpuncel replied:
after looking at it, it's probably not worth it because you end up with a nested lambda situation (one calls `shouldRetryReset`, one calls `shouldRetryHedgedPerTryTimeout` and both need to provide the `doRetry()` lambda)

mpuncel commented:
just merged master again and addressed PR feedback. I also got a screenshot of tracing showing hedging working for a real gRPC request at Square.

It shows an app calling itself with a 100ms per try timeout and the server takes longer than that. It looks a bit odd because the 2 spans have the same parent but don't start at the same time, and then that one server span is long because cancellation wasn't handled (it just does a `sleep`).

<img width="1254" alt="Screen Shot 2019-05-22 at 1 55 51 PM" src="https://user-images.githubusercontent.com/1444855/58212151-de311d00-7cbb-11e9-9b4f-e668ddbc9383.png">

mattklein123 commented:
Looks like CI issues. Note also that docs is broken on master right now so watch for a fix for that.

/wait
